Refactor description generator and make views' description usable in particles' ui

mmandlis opened the pull request:
Scott, 
I'll need to update and make some changes to merge with your recent PR.
Meantime, could you, please, glance and how i did the plumbing between views descriptions in particles and slot-composer and arc... Seems a bit hacky, but i couldn't think of a better way at the moment.

mmandlis commented:
FYI I merged with your recent changes

sjmiles commented:
Will study. :)

sjmiles commented on runtime/browser/demo/demo.js:
```diff
@@ -78,11 +78,13 @@ class DemoFlow extends DemoBase {
-      slotComposer: new SlotComposer(this._root.querySelector('[particle-container]'))
+      slotComposer: slotComposer
     });
+    slotComposer.arc = arc;
```
ContextFactory should probably do this assignment.

sjmiles commented on particles/Chooser/Chooser.js:
```diff
@@ -102,7 +102,7 @@ ${productStyles}
     _render(props, state) {
       return {
-        person: 'Claire',
+        choices: '{{choices.description}}',
```
Mustache (`{{...}}`) expression replacing is already a core capability of the templating engine, I'd like to see if we can leverage that system instead of recreating it. If we can come up with a way of getting the description information into the Particle as a data-model, then the Particle can include that data into the render-model (return value of `_render()`) and let the template engine interpolate the content.

mmandlis replied:
The reason I did it this way was to avoid sending view description provided by one particle to another, because it can expose undesired information (like here, the Chooser just shows a list of products, but it shouldn't know that these come from Claire's wishilist). I want to keep the descriptions on the outer pec side.
I mentioned it in the design doc in the "Using View description in Particles" section.

Let me know what you think.

sjmiles commented on runtime/dom-slot.js:
```diff
@@ -46,13 +46,13 @@ class DomSlot extends Slot {
   }
   // TODO(sjmiles): SlotManager calls here
-  render(content, eventHandler) {
+  render(content, arc, eventHandler) {
```
Plumbing the arc through feels like too-tight coupling between the objects. I suspect we can avoid this as described above.

sjmiles commented on runtime/dom-slot.js:
```diff
@@ -73,13 +73,27 @@ class DomSlot extends Slot {
+      if (matchers && matchers.length > 1) {
+        let view = arc.findViewByName(this.particleSpec.particle.name, matchers[1]);
+        if (view && view.description)
+          model[key] = view.description;
```
If the description is already attached to the view, couldn't the particle just do (simplified):
```js
_render(props, state) {
  return {
    choices: this.views.get('choices').description
  }
}
```  

?

mmandlis commented:
Addressed comments, as discussed offline - I ended up not needing the Arc at all in slot-composer and dom-slot, so the code is much cleaner than it used to be.
